crates/client/block_import/src/verify_apply.rs

```diff
@@ -0,0 +1,281 @@
+
+use dc_db::{DeoxysBackend, DeoxysStorageError};
+use dp_block::{
+    header::PendingHeader, BlockId, BlockTag, DeoxysBlockInfo, DeoxysBlockInner, DeoxysMaybePendingBlock,
+    DeoxysMaybePendingBlockInfo, DeoxysPendingBlockInfo, Header,
+};
+use dp_convert::ToFelt;
+use starknet_api::core::ChainId;
+use starknet_core::types::Felt;
+use starknet_types_core::hash::{Poseidon, StarkHash};
+
+use crate::{
+    BlockImportError, BlockImportResult, PendingBlockImportResult, PreValidatedBlock, PreValidatedPendingBlock,
+    RayonPool, UnverifiedHeader, ValidatedCommitments, Validation,
+};
+
+mod classes;
+mod contracts;
+
+pub struct VerifyApply {
+    pool: Arc<RayonPool>,
+    backend: Arc<DeoxysBackend>,
+    // Only one thread at once can verify_apply.
+    mutex: tokio::sync::Mutex<()>,
+}
+
+impl VerifyApply {
+    pub fn new(backend: Arc<DeoxysBackend>, pool: Arc<RayonPool>) -> Self {
+        Self { pool, backend, mutex: Default::default() }
+    }
+
+    /// This function wraps the [`verify_apply_inner`] step, which runs on the rayon pool, in a tokio-friendly future.
+    pub async fn verify_apply(
+        &self,
+        block: PreValidatedBlock,
+        validation: Validation,
+    ) -> Result<BlockImportResult, BlockImportError> {
+        let _exclusive = self.mutex.lock();
+
+        let backend = Arc::clone(&self.backend);
+        self.pool.spawn_rayon_task(move || verify_apply_inner(&backend, block, validation)).await
+    }
+
+    /// See [`Self::verify_apply`].
+    pub async fn verify_apply_pending(
+        &self,
+        block: PreValidatedPendingBlock,
+        validation: Validation,
+    ) -> Result<PendingBlockImportResult, BlockImportError> {
+        let _exclusive = self.mutex.lock();
+
+        let backend = Arc::clone(&self.backend);
+        self.pool.spawn_rayon_task(move || verify_apply_pending_inner(&backend, block, validation)).await
+    }
+}
+
+/// This needs to be called sequentially, it will apply the state diff to the db, verify the state root and save the block.
+/// This runs on the [`rayon`] threadpool however as it uses parallelism inside.
+pub fn verify_apply_inner(
+    backend: &DeoxysBackend,
+    block: PreValidatedBlock,
+    validation: Validation,
+) -> Result<BlockImportResult, BlockImportError> {
+    // Check block number and block hash against db
+    let (block_number, parent_block_hash) =
+        check_parent_hash_and_num(backend, block.header.parent_block_hash, block.unverified_block_number)?;
+
+    // Update contract and its storage tries
+    let global_state_root = update_tries(backend, &block, block_number)?;
+
+    // Block hash
+    let (block_hash, header) = block_hash(&block, &validation, block_number, parent_block_hash, global_state_root)?;
+
+    // store block, also uses rayon heavily internally
+    backend
+        .store_block(
+            DeoxysMaybePendingBlock {
+                info: DeoxysMaybePendingBlockInfo::NotPending(DeoxysBlockInfo {
```

Comment:
`NotPending` might not be the best name, but rather `Closed` or `Build` wdyt ?

Reply:
i've been thinking about this for some time, in starknet-types-core they say MaybePendingBlockInfo::Block/Pending and i kind of don't like it
I really like Closed but I don't know if it's that much better

Reply:
Definitely is